cardano-cli/src/Cardano/CLI/Read.hs

```diff
@@ -518,21 +518,21 @@ readFileTx file = do
     Left e -> fmap unCddlTx <$> acceptTxCDDLSerialisation file e
     Right tx -> return $ Right tx
 
--- IncompleteCddlFormattedTx is an CDDL formatted tx or partial tx
--- (respectively needs additional witnesses or totally unwitnessed)
--- while UnwitnessedCliFormattedTxBody is CLI formatted TxBody and
--- needs to be key witnessed.
+newtype IncompleteCddlTxBody =
+  IncompleteCddlTxBody { unIncompleteCddlTxBody :: InAnyShelleyBasedEra TxBody }
 
-data IncompleteTx
-  = UnwitnessedCliFormattedTxBody (InAnyShelleyBasedEra TxBody)
-  | IncompleteCddlFormattedTx (InAnyShelleyBasedEra Tx)
-
-readFileTxBody :: FileOrPipe -> IO (Either CddlError IncompleteTx)
+readFileTxBody :: FileOrPipe -> IO (Either CddlError IncompleteCddlTxBody)
 readFileTxBody file = do
   eTxBody <- readFileInAnyShelleyBasedEra AsTxBody file
   case eTxBody of
-    Left e -> fmap (IncompleteCddlFormattedTx . unCddlTx) <$> acceptTxCDDLSerialisation file e
-    Right txBody -> return $ Right $ UnwitnessedCliFormattedTxBody txBody
+    Left e -> do
+      cddlErrOrTx <- acceptTxCDDLSerialisation file e
+      case cddlErrOrTx of
+        Left cddlErr -> return $ Left cddlErr
+        Right cddlTx -> do
+          InAnyShelleyBasedEra sbe tx <- pure $ unCddlTx cddlTx
+          return $ Right $ IncompleteCddlTxBody $ inAnyShelleyBasedEra sbe $ getTxBody tx
+    Right txBody -> return $ Right $ IncompleteCddlTxBody txBody
```

Comment:
@Jimbo4350> so you want to treat those two cases in the same manner, right!?

Reply:
You're almost there :+1: 

So we have: 
```
acceptTxCDDLSerialisation
  :: FileOrPipe
  -> FileError TextEnvelopeError
  -> IO (Either CddlError CddlTx)
acceptTxCDDLSerialisation file err =
  case err of
   e@(FileError _ (TextEnvelopeDecodeError _)) ->
      first (CddlErrorTextEnv e) <$> readCddlTx file
   e@(FileError _ (TextEnvelopeAesonDecodeError _)) ->
      first (CddlErrorTextEnv e) <$> readCddlTx file
   e@(FileError _ (TextEnvelopeTypeError _ _)) ->
      first (CddlErrorTextEnv e) <$> readCddlTx file
   e@FileErrorTempFile{} -> return . Left $ CddlIOError e
   e@FileDoesNotExistError{} -> return . Left $ CddlIOError e
   e@FileIOError{} -> return . Left $ CddlIOError e
```
We cased on the `TextEnvelope` related errors because by default we used the deprecated "intermediate cli" format. Because it's no longer possible to create that format, and we will no longer support it, we can remove `acceptTxCDDLSerialisation` entirely and simply rely on the `TextEnvelope` encoding because we now _default_ to the ledger's CDDL format.

Reply:
See below:
```
instance IsShelleyBasedEra era => SerialiseAsCBOR (Tx era) where
    serialiseToCBOR (ShelleyTx sbe tx) =
      shelleyBasedEraConstraints sbe $ serialiseShelleyBasedTx tx

    deserialiseFromCBOR _ bs =
      shelleyBasedEraConstraints (shelleyBasedEra :: ShelleyBasedEra era)
        $ deserialiseShelleyBasedTx (ShelleyTx shelleyBasedEra) bs

-- | The serialisation format for the different Shelley-based eras are not the
-- same, but they can be handled generally with one overloaded implementation.
--
serialiseShelleyBasedTx :: forall ledgerera .
                           L.EraTx ledgerera
                        => L.Tx ledgerera
                        -> ByteString
serialiseShelleyBasedTx = Plain.serialize'
```

We call ledger's `serialize'` and `TextEnvelope` is serialized as follows:
```
class SerialiseAsCBOR a => HasTextEnvelope a where
    textEnvelopeType :: AsType a -> TextEnvelopeType

    textEnvelopeDefaultDescr :: a -> TextEnvelopeDescr
    textEnvelopeDefaultDescr _ = ""

serialiseToTextEnvelope :: forall a. HasTextEnvelope a
                        => Maybe TextEnvelopeDescr -> a -> TextEnvelope
serialiseToTextEnvelope mbDescr a =
    TextEnvelope {
      teType    = textEnvelopeType ttoken
    , teDescription   = fromMaybe (textEnvelopeDefaultDescr a) mbDescr
    , teRawCBOR = serialiseToCBOR a
    }
  where
    ttoken :: AsType a
    ttoken = proxyToAsType Proxy
```
Hope this makes things clearer!